### ISSUE TITLE
[Performance] Remove instances of `torch.nonzero()` where appropriate

### ISSUE THREAD

**Issue by vincentzed** [sgl-project/sglang#9889]
Right now we have:
```
4 results - 4 files

python/sglang/srt/configs/janus_pro.py:
  478          image_token_mask: torch.Tensor = (input_ids == self.image_id).to(torch.bool)
  479:         image_indices = image_token_mask.nonzero()
  480          input_ids, num_image_tokens = self.add_image_token(

python/sglang/srt/models/gemma3_mm.py:
  211          # Distinguish sequences by position id 0
  212:         start_indices = (positions == 0).cpu().nonzero()
  213          num_seqs = len(start_indices)

python/sglang/srt/models/phi4mm_utils.py:
  83      seq_range = torch.arange(0, x_len).unsqueeze(-1)  # seq_range size: [x_len, 1]
  84:     idx = ((seq_range < end_pad) & (seq_range >= start_pad)).nonzero()[
  85          :, 1
```
We should explore whether we can update these to reduce the host-device sync. You could profile it before trying to update to see if it really happens, or analyze the call sites deeply and test changes.
Why? See below.

> **Warning**
> When input is on CUDA, [torch.nonzero()](https://docs.pytorch.org/docs/2.8/generated/torch.nonzero.html#torch.nonzero) causes host-device synchronization.

**Comment by zzhbrr** [comment]
Hi. 
Regarding the code in `janus_pro.py` and `phi4mm_utils.py`, I actually performed inference and found that the tensors on which the torch.nonzero()  is executed are all on the CPU. So there should be no synchronization issue as mentioned earlier. 
As for the code in `gemma3_mm.py`, torch.nonzero() appears in the `prepare_attn_masks()` function, but I could not find any references to this function in the codebase. I think this function is unused. 
I think none of the three instances where torch.nonzero() appears will introduce host-device synchronization. 
I’m wondering if my understanding is correct.

**Comment by vincentzed** [comment]
> Hi. Regarding the code in `janus_pro.py` and `phi4mm_utils.py`, I actually performed inference and found that the tensors on which the torch.nonzero() is executed are all on the CPU. So there should be no synchronization issue as mentioned earlier. As for the code in `gemma3_mm.py`, torch.nonzero() appears in the `prepare_attn_masks()` function, but I could not find any references to this function in the codebase. I think this function is unused. I think none of the three instances where torch.nonzero() appears will introduce host-device synchronization. I’m wondering if my understanding is correct.

sounds awesome! what approach did you use with the profiling tool? then you can be certain. otherwise, happy to hear about how you did it. i also noticed the unused function. i think that was weird but not really rare in this codebase, so it could probably be removed or someone could run `blame` and trace back and see why it was introduced.

**Comment by zzhbrr** [comment]
@vincentzed.
I did not use any profiling tools. Instead, I performed single-step debugging in VSCode while executing actual inference tasks and observed that all tensors executing torch.nonzero() were on the CPU. I am unsure whether this approach is correct.

Regarding the unused function `prepare_attn_masks()`: This code was introduced in https://github.com/sgl-project/sglang/pull/4424 and was adapted from the vLLM codebase https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/gemma3_mm.py#L637 . The function is used to generate attention masks because **the image tokens in the Gemma-3 model require bidirectional attention**:
*  In vLLM v0, this function was used to set the attention mask, achieving fully correct attention, but it used torch native attention implementation.
* In vLLM v1, this function was not used, resulting in incorrect attention (i.e., causal attention is still applied between image tokens). However, this allows the use of a more efficient attention backend without introducing significant errors, as illustrated in https://github.com/vllm-project/vllm/pull/14660
* Similarly, in SGLang, this function is also unused — meaning “incorrect” attention is implemented — likely because some attention backends do not support custom masks during the extend phase.
I think we can remove this function and add some comments in the code. For attention backends that support custom masks, we can implement correct attention.

**Comment by vincentzed** [comment]
> [@vincentzed](https://github.com/vincentzed). I did not use any profiling tools. Instead, I performed single-step debugging in VSCode while executing actual inference tasks and observed that all tensors executing torch.nonzero() were on the CPU. I am unsure whether this approach is correct.
> 
> Regarding the unused function `prepare_attn_masks()`: This code was introduced in [#4424](https://github.com/sgl-project/sglang/pull/4424) and was adapted from the vLLM codebase https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/gemma3_mm.py#L637 . The function is used to generate attention masks because **the image tokens in the Gemma-3 model require bidirectional attention**:
> 
>     * In vLLM v0, this function was used to set the attention mask, achieving fully correct attention, but it used torch native attention implementation.
> 
>     * In vLLM v1, this function was not used, resulting in incorrect attention (i.e., causal attention is still applied between image tokens). However, this allows the use of a more efficient attention backend without introducing significant errors, as illustrated in [[Model] Add support for Gemma 3 vllm-project/vllm#14660](https://github.com/vllm-project/vllm/pull/14660)
> 
>     * Similarly, in SGLang, this function is also unused — meaning “incorrect” attention is implemented — likely because some attention backends do not support custom masks during the extend phase.
>       I think we can remove this function and add some comments in the code. For attention backends that support custom masks, we can implement correct attention.

Sounds good, should be ok. You can probably open an issue about the existing problem, or you can fix it directly.

**Comment by zzhbrr** [comment]
> Sounds good, should be ok. You can probably open an issue about the existing problem, or you can fix it directly.

ok. I'll try to fix it.

**Comment by zzhbrr** [comment]
> Hello [@zzhbrr](https://github.com/zzhbrr), are you working on this?

Hi. I have submitted a pull request, but sadly, it hasn't been merged yet.

 https://github.com/sgl-project/sglang/issues/10309, https://github.com/sgl-project/sglang/pull/10707

**Comment by yinli-systems** [comment]
I re-checked this against current `main` and the issue looks mostly stale now.

- The original Janus/Phi/Gemma comments above still seem accurate: those nonzero sites are CPU-side or were not the hot runtime sync path that this issue originally suspected.
- The Gemma-3 bidirectional image-attention path mentioned in this thread was later implemented/merged in #10707, and `prepare_attn_masks` is now called from the Gemma model path on `main`.
- #19635 is still open, but the maintainer feedback there also says the touched nonzero calls are not currently causing H2D synchronization, so the performance motivation for that specific patch is weak.

There are still a few `.nonzero(as_tuple=True)` calls in runtime code (for example some logits/Mamba metadata paths), but they are tied to downstream boolean indexing or variable-length gather/scatter semantics. I do not think they should be changed mechanically without profiling a concrete GPU sync and preserving those shape/order semantics.
